sklearn/tests/test_impute.py

```diff
@@ -1030,6 +1033,591 @@ def test_missing_indicator_sparse_param(arr_type, missing_values,
             assert isinstance(X_trans_mask, np.ndarray)
 
 
+def test_knn_imputer_shape():
+    # Verify the shapes of the imputed matrix for different weights and
+    # number of neighbors.
+    n_rows = 10
+    n_cols = 2
+    X = np.random.rand(n_rows, n_cols)
+    X[0, 0] = np.nan
+
+    for weights in ['uniform', 'distance']:
+        for n_neighbors in range(1, 6):
+            imputer = KNNImputer(n_neighbors=n_neighbors, weights=weights)
+            X_imputed = imputer.fit_transform(X)
+            assert X_imputed.shape == (n_rows, n_cols)
+
+
+def test_knn_imputer_zero_missing_value():
+    # Test imputation when missing_values == 0
+    missing_values = 0
+    n_neighbors = 2
+    imputer = KNNImputer(missing_values=missing_values,
+                         n_neighbors=n_neighbors,
+                         weights="uniform")
+
+    # Test with missing_values=0 when NaN present
+    X = np.array([
+        [np.nan, 0, 0, 0, 5],
+        [np.nan, 1, 0, np.nan, 3],
+        [np.nan, 2, 0, 0, 0],
+        [np.nan, 6, 0, 5, 13],
+    ])
+    msg = (r"Input contains NaN, infinity or a value too large for "
+           r"dtype\('float64'\)")
+    with pytest.raises(ValueError, match=msg):
+        imputer.fit(X)
+
+    # Test with % zeros in samples > sample_max_missing
+    X = np.array([
+        [1, 0, 0, 0, 5],
+        [2, 1, 0, 2, 3],
+        [3, 2, 0, 0, 0],
+        [4, 6, 0, 5, 13],
+    ])
+    msg = "Some columns have more than {}% missing values".format(
+        imputer.sample_max_missing * 100)
+    with pytest.raises(ValueError, match=msg):
+        imputer.fit(X)
+
+
+def test_knn_imputer_zero_nan_missing_values():
+    # Test with an imputable matrix and also compare with missing_values=np.NaN
+    X_zero = np.array([
+        [1, 0, 1, 1, 1.],
+        [2, 2, 2, 2, 2],
+        [3, 3, 3, 3, 0],
+        [6, 6, 0, 6, 6],
+    ])
+
+    X_nan = np.array([
+        [1, np.nan, 1,      1,      1.],
```

Comment:
I'm not sure how I feel about this kind of formatting.

It's clearly visually nicer, but it's a pain as soon as you need to change something as you have to adjust the white spaces.

Pep8 discourages the use of a similar pattern:

```
Yes:

x = 1
y = 2
long_variable = 3

No:

x             = 1
y             = 2
long_variable = 3

```

Let's wait the other's feedback